Transformations: improve the reduce transformer

ryantxu opened the pull request:
The current `reduce` transformer was designed to have output that matches the old angular table format; this structure is hard to use when using it for other types of visualization and in particular it loses all label information.

This PR adds an additional mode to the reduce transformer that just operations on the field:

![image](https://user-images.githubusercontent.com/705951/94498159-519b4680-01ae-11eb-8206-8b6a37029b01.png)

![image](https://user-images.githubusercontent.com/705951/94498182-6081f900-01ae-11eb-92db-ece174ad940a.png)

TODO:
- [x] testing
- [ ] docs

ryantxu committed:
improve reduce transformer

ryantxu commented:
cc: @mikempx

torkelo commented:
While your looking at this reducer, would be interesting to explore how we can solve the series to rows scenario but keep the time series values in a single field value (so we can have a sparkline cell mode in the table). So

kylebrandt commented:
The one result per frame is how GEL currently does things:

![image](https://user-images.githubusercontent.com/1692624/94597232-4fdb8c80-025b-11eb-9193-8bf744055d47.png)

I think it is interesting to note that I think both the one dataframe per thing, and the one field per thing in a single frame, don't always visualize well (in the single frame, case with too many columns).

In terms of visualization, I think it is best to take the above and pass it through what I think "Labels to Fields" does:

![image](https://user-images.githubusercontent.com/1692624/94597463-98934580-025b-11eb-86c0-4083e68ff429.png)

(also might look nice as a 90 degree rotated bar graph that you can scroll down, sorted by the value.)

ryantxu committed:
merge master

ryantxu committed:
set values

torkelo commented on packages/grafana-data/src/transformations/transformers/reduce.ts:
```diff
@@ -33,14 +37,55 @@ export const reduceTransformer: DataTransformerInfo<ReduceTransformerOptions> =
 
         const processed: DataFrame[] = [];
+        // Collapse all matching fields into a single row
+        if (options.mode === ReduceTransformerMode.ReduceFields) {
+          for (const series of data) {
+            const fields: Field[] = [];
+            for (const field of series.fields) {
+              if (matcher(field, series, data)) {
+                const results = reduceField({
+                  field,
+                  reducers,
+                });
+                for (const reducer of reducers) {
+                  const value = results[reducer];
+                  const copy = {
+                    ...field,
+                    values: new ArrayVector([value]),
+                  };
+                  copy.state = undefined;
+                  if (reducers.length > 1) {
+                    if (!copy.labels) {
+                      copy.labels = {};
+                    }
+                    copy.labels['reducer'] = fieldReducers.get(reducer).name;
+                  }
+                  fields.push(copy);
+                }
+              }
+            }
+            if (fields.length) {
+              processed.push({
+                ...series,
+                fields,
+                length: 1, // always one row
+              });
+            }
+          }
+
+          return processed;
+        }
```
This reducer function is 120 lines of pretty dense code. Would be nice to be if we could break it up a bit. And nitpick: I struggle reading large code blocks like this that has no breaks between control statement, it all mixes and together and I mistake statements to look like they form a single expression but are really completely separate expressions.

torkelo reviewed (changes requested):
Looks good! Think this reducer needs a minor refactoring (extract method), to make it a bit more readable. And this new logic needs some tests :)

ryantxu committed:
split into two functions

ryantxu committed:
add tests

ryantxu commented:
@torkelo, I split this into two smaller functions and added a simple test.  I will still add docs before I hope to merge, but check if the implementation looks better.

torkelo reviewed (approved):
Yes, a lot easier to navigate and read the code now!
